### PR TITLE
Refactor BuildConfig controller to use Informers

### PR DESCRIPTION
**Pull request by csrwng**
Switches to using informers for the BuildConfig controller.
Switches to using caches for Build pruning.

Fixes https://github.com/openshift/origin/issues/14049
Depends on https://github.com/openshift/origin/pull/14289

### CODE REVIEW

**Comment by csrwng** [comment]
@openshift/devex ptal

**Comment by bparees** [comment]
@csrwng so this fixes https://github.com/openshift/origin/issues/14049 ?

File: pkg/build/controller/buildconfig/buildconfig_controller.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+	buildConfigStoreSynced func() bool
+
+	recorder record.EventRecorder
+}
+
+func NewBuildConfigController(openshiftClient osclient.Interface, kubeExternalClient kexternalclientset.Interface, buildConfigInformer shared.BuildConfigInformer, buildInformer shared.BuildInformer) *BuildConfigController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeExternalClient.Core().RESTClient()).Events("")})
+
+	buildClient := buildclient.NewOSClientBuildClient(openshiftClient)
```
Comment (bparees) [inline pkg/build/controller/buildconfig/buildconfig_controller.go]: so this one is not using cached builds?

**Comment by bparees** [comment]
i guess it doesn't fix all of #14049, just the buildconfig side.  the build controller logic that calls it would still need to change, unless that's already covered by your changes to the build controller.

File: pkg/build/controller/buildconfig/buildconfig_controller.go (hunk 1 above)
Comment (csrwng) [inline pkg/build/controller/buildconfig/buildconfig_controller.go]: it's not ... it's used for the patch and delete operations

**Comment by csrwng** [comment]
> @csrwng so this fixes #14049 ?

yes

**Comment by csrwng** [comment]
> i guess it doesn't fix all of #14049, just the buildconfig side. the build controller logic that calls it would still need to change, unless that's already covered by your changes to the build controller.

This PR includes changes to the build controller to use the informer for pruning

**Comment by bparees** [comment]
[testextended][extended:core(builds)]

**Comment by bparees** [comment]
lgtm if extended tests pass

**Comment by bparees** [comment]
[test]

**Comment by csrwng** [comment]
https://github.com/openshift/origin/issues/14229
[test]

**Comment by bparees** [comment]
the failed extended test is for build pruning so that looks suspicious given the changes in this pr.

**Comment by csrwng** [comment]
> the failed extended test is for build pruning so that looks suspicious given the changes in this pr.

The test passes locally. Will look more closely at the logs from the test and see if there's some flake introduced by the use of the cache.

**Comment by csrwng** [comment]
@bparees so using the cache, it looks like the build that was just updated to completed is not updated in the cache yet when we list the builds for pruning. Would you be ok with reverting the use of the cache for the build lister?

**Comment by csrwng** [comment]
@bparees added to commit to revert to using the rest client to list builds for pruning, ptal

**Comment by bparees** [comment]
lgtm

**Comment by bparees** [comment]
woo. squish+merge

**Comment by csrwng** [comment]
[merge]

**Comment by bparees** [comment]
acceptable extended test flake.

**Comment by jim-minter** [comment]
@csrwng is there an issue or a comment on an issue somewhere, so that we can remember to come back to the pruning cache thing?

**Comment by csrwng** [comment]
@jim-minter we do now :) https://github.com/openshift/origin/issues/14638
thx

**Comment by smarterclayton** [comment]
[test]

**Comment by csrwng** [comment]
Flake https://github.com/openshift/origin/issues/13980

**Pull request merged**